enterprise/backend/src/metabase_enterprise/data_editing/coerce.clj

```diff
@@ -0,0 +1,50 @@
+(ns metabase-enterprise.data-editing.coerce
+  (:import
+   (clojure.lang BigInt)
+   (java.time ZonedDateTime)
+   (java.time.format DateTimeFormatter)))
+
+(set! *warn-on-reflection* true)
+
+(defn- json-zdt->unix-millis ^long [s]
+  (-> s ZonedDateTime/parse .toInstant .toEpochMilli))
+
+(defn- json-zdt->unix-seconds [s]
+  (-> s json-zdt->unix-millis (quot 1000)))
+
+(defn- json-zdt->unix-micros [s]
+  (-> s json-zdt->unix-millis ^BigInt (* 1000N) .toBigInteger))
+
+(defn- json-zdt->unix-nanos [s]
+  (-> s json-zdt->unix-millis ^BigInt (* 1000000N) .toBigInteger))
+
+(let [formatter (DateTimeFormatter/ofPattern "yyyyMMddHHmmSS")]
+  (defn- json-zdt->yyyymmddhhmmss [s]
+    (-> s ZonedDateTime/parse (.format formatter))))
+
+;; the date/time coercions are highly ambigious
+;; in one direction one can make a decision (take the local date, or clock time)
+;; but how do you reverse it?
+;; Often the original SQL string might have been a full date, a zoned or unzoned ISO string,
+;; Needs to be considered properly at some later time.
+;; Idea: make the coercion dependent on previous database string, so if lossy - modify *just* the time, or the date component
```

Comment:
That's a bit magical for my liking! We also don't know whether this "previous" value is from a "concurrent" change (anything since the editing user opened the table)